### PR TITLE
Container revamp

### LINKED ISSUE #1099: Proposal to replace JuMPContainers
**Updated** Here's a proposal to completely replace JuMPContainer.
The ``@variable`` macro can return three types of containers:

1. ``@variable(m, l[i] <= x[i=1:N,j=1:M] <= u[i])`` returns an ``Array`` as it does now. More generally, this applies whenever all dimensions are one-based unit-step intervals. ``I = 1:N; @variable(..., x[I])`` falls under this case also.

2. ``@variable(m, l[i] <= x[i=A,j=B] <= u[i])`` returns an AxisArray as in https://github.com/rdeits/AxisArrayVariables.jl. Index sets are restricted to those [supported by AxisArrays](https://github.com/JuliaArrays/AxisArrays.jl#indexing-axes).

3. ``@variable(m, l[i] <= x[i=1:N,j=i:K ; i != j] <= u[i])`` returns a plain Julia `Dict` *indexed by explicit tuples*, e.g., ``x[(i,j)]``  (but ``x[i,j]`` also works, as @tkoolen noted). This case supports conditions, triangular indexing, and arbitrary index sets. If conditions or triangular indexing is detected at compile time, we choose this option. I don't think we can do a good job detecting index sets not supported by AxisArrays, so for this case I'd propose...

**Manual specification of container type**: ``@variable(..., container = [Array|AxisArray|Dict])``. Force one of the three container types and error if not supported.

Someone currently using syntax of the form (2) but with a set that's not supported by AxisArrays would now have to specify ``container = Dict``.

On the plus side we completely resolve all issues with JuMPContainers (https://github.com/JuliaOpt/JuMP.jl/issues/1047). JuMPDict vs JuMPArray has never been a clear point for users, and using fake multi-dimensional indexing for what's really a dictionary leads to mismatched expectations because we don't support slicing or other array-like operations for JuMPDicts.

I'm a bit worried about the dependency on AxisArrays since I haven't used it much, and it could open JuMP to upstream breakage. I'd also like to hear opinions on how this affects JuMP's ease of use for new users.

@rdeits @tkoolen @davidanthoff @chkwon

### PR DESCRIPTION
**Pull request by mlubin**
Replace JuMPDict with Dict. Rewrite JuMPArray to be compatible with AbstractArray. Explicit keyword argument in macro to force container type.

Unfortunately we can't use AxisArrays at the moment because of JuliaArrays/AxisArrays.jl#117. I'm hoping that we can swap AxisArray for JuMPArray before JuMP 1.0/0.19, but for now I'd like to go ahead with this since the basic functionality is there. If we keep JuMPArray, it will need to be polished and performance-tested before the next release.

Broadcast is implemented for JuMPArray, so we can just do `resultvalue.(A)` when `A` is an Array or a JuMPArray (but not when it's a Dict).

See commit for list of 11 issues that this PR closes (once merged into master).

Remaining to do for merge:
- [x] Add `container` keyword argument to all macros
- [x] Go through issues listed in #1047 and check if they are addressed
- [x] Better test coverage

### CODE REVIEW

**Comment by mlubin** [comment]
I plan on adding some more tests but otherwise this is ready for review.

File: src/JuMPArray.jl [hunk 1]
```diff
@@ -3,96 +3,214 @@
...
 end
+Base.getindex(A::JuMPArray, idx::CartesianIndex) = A.data[idx]
+
+Base.setindex!(A::JuMPArray, v, idx...) = A.data[to_index(A,idx...)...] = v
+Base.setindex!(A::JuMPArray, v, idx::CartesianIndex) = A.data[idx] = v
+
+# AbstractArray interface
+
+Base.linearindices(A::JuMPArray) = error("JuMPArray does not support this operation.")
+Base.size(A::JuMPArray) = error("JuMPArray does not define this operation")
```
[the remark this answers is not in the document]
Comment (mlubin) [reply src/JuMPArray.jl]: We don't define `size` for the old JuMPArray, and neither does OffsetArrays. Defining it causes `end` to return the wrong result.

File: src/containers.jl [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+used to determine the appropriate container:
+
+1. If all index sets are either explicit `Expr(:(:), 1, B)` objects for any `B` or symbols which refer to objects of type `Base.OneTo`, then an `Array` is generated of the appropriate size. Types of symbols/expressions are not known at compile time, so we defer to type-safe functions to check the `Base.OneTo` condition.
+
+2. If condition (1) does not hold, and the index sets are independent (the index variable for one set does not appear in the definition of another), then an `JuMPArray` is generated of the appropriate size.
+
+3. Otherwise, generate an empty `Dict{Any,T}`.
+
+The second element of the return tuple is a `Bool`, `true` if the container type
+automatically checks for duplicate terms in the index sets and `false` otherwise.
```
[the remark this answers is not in the document]
Comment (mlubin) [reply src/containers.jl]: The issue is that it depends on the processing of `indexvars` and `indexsets` (`hasdependentsets`). I agree that this solution isn't too tidy though.

**Comment by mlubin** [comment]
> Do you have a sense of the performance relative to the current implementation?

There's no reason to benchmark `JuMPArray` right now because it will just be slow. When rewritten with generated functions it should have similar performance, except no special handling for interval sets (but this could be added transparently without breaking the API).

**Comment by joehuchette** [comment]
OK. What's the roadmap then with this PR, generated functions, and AxisArrays?

**Comment by mlubin** [comment]
> What's the roadmap then with this PR, generated functions, and AxisArrays?

I'd like to merge this PR soon and get back to MOI work. When MOI is close to being wrapped up we can revisit and see if AxisArrays has implemented the functionality we need or if we need to tidy up JuMPArray.

**Pull request merged**